racket/src/ChezScheme/c/version.h

```diff
@@ -326,7 +326,11 @@ typedef int tputsputcchar;
 #endif
 #if defined(__arm64__)
 # define S_MAP_CODE  MAP_JIT
```

Comment:
Does moving this to the non-`TARGET_OS_IPHONE` branch below help?

Reply:
Yup, that does help get past the allocation problem! Now it aborts because of a normal return from `S_generic_invoke`:

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001086a0150 in 0x1086a0150 ()
#1	0x00000001022aa414 in S_call_help ()
#2	0x00000001022a904c in boot_call at /Users/bogdan/sandbox/racket-ios/racket/src/build/cs/c/ChezScheme/tarm64osx/c/scheme.c:353
#3	0x00000001022a8f38 in load at /Users/bogdan/sandbox/racket-ios/racket/src/build/cs/c/ChezScheme/tarm64osx/c/scheme.c:903
#4	0x00000001022a8c00 in Sbuild_heap at /Users/bogdan/sandbox/racket-ios/racket/src/build/cs/c/ChezScheme/tarm64osx/c/scheme.c:1160
#5	0x00000001022b305c in racket_boot at /Users/bogdan/sandbox/racket-ios/racket/src/cs/c/boot.c:156
#6	0x00000001021ee088 in racket_init at /Users/bogdan/sandbox/rbggame/shared/game.c:19
#7	0x00000001021eefe4 in AppDelegate.application(_:didFinishLaunchingWithOptions:) at /Users/bogdan/sandbox/rbggame/ios/rbggame/rbggame/AppDelegate.swift:21
#8	0x00000001021efff8 in @objc AppDelegate.application(_:didFinishLaunchingWithOptions:) ()
#9	0x00000001acc92540 in -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] ()
#10	0x00000001acc945a8 in -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] ()
#11	0x00000001acc9a004 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#12	0x00000001ac2f1950 in -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] ()
#13	0x00000001acc95db0 in -[UIApplication _compellApplicationLaunchToCompleteUnconditionally] ()
#14	0x00000001acc961dc in -[UIApplication _run] ()
#15	0x00000001acc9ba74 in UIApplicationMain ()
#16	0x00000001bdb2d5bc in UIApplicationMain(_:_:_:_:) ()
#17	0x00000001021f07d8 in static UIApplicationDelegate.main() ()
#18	0x00000001021f0750 in static AppDelegate.$main() ()
#19	0x00000001021f081c in main ()
#20	0x00000001a9f1d6c0 in start ()
```

I'll be able to dig into this a little more on Saturday.

Reply:
The problem here was actually just that the code segments weren't executable. Explicitly flipping between W and X in 56916d6459dd96378882c8df8a3d2c4e16d54fd6 gets things much farther. `petite.boot` and `scheme.boot` both boot up properly now, but I get an issue when loading `racket.boot`:

```
fasl-read: incompatible fasl-object machine-type 'ta6osx found in #<binary input port bytevector>
```

so something about the build must still be incorrect.

Reply:
Ah, actually, the issue is with how I'm generating the C module from my Racket code.  Racket itself boots up properly after the changes in https://github.com/racket/racket/commit/56916d6459dd96378882c8df8a3d2c4e16d54fd6 ! 

I'm using this command to generate a C mod from my Racket code:

   racket -C -G ~/sandbox/racket-ios/racket/etc/ -X ~/sandbox/racket-ios/racket/collects -l- raco ctool --c-mods shared/embed.c core/main.rkt

Unless I'm doing something wrong, there must be some kind of bug in `cext-lib` that's causing it to produce output for the source machine.

Reply:
For cross-compilation to work with CS, cross-compiler "patch" files need to be specified through the `--cross-compiler` argument to `racket`. I see that this is not explained in the documentation, and it's really not worked out fully; its there just enough to cross-build distributions, right now.

For a distribution build, the second argument to `--cross-compiler` points to "racket/src/build/cs/c". I think the two files needed from there will be "compile-xpatch.tarm64osx" and "library-xpatch.tarm64osx", so those could be copied to a different directory to be given to `--cross-compilers`. Probably, the `make install` step should copy the patch files to `racket/lib`.

Reply:
Swapping pages between `PROT_WRITE` and `PROT_EXEC` is not going to be a good long-term solution. For one thing, it's slow. More importantly, it can't work right with threads, because one thread might be trying to run some (immobile) code while another thread needs to write on the same pages.

My vague impression ws that W&X is allowed on iOS with the right entitlements and maybe flags like `MAP_JIT`. Is that just not right?

Reply:
On further reflection, I'm not clear on why `raco ctool` didn't complain about a missing `tarm64osx` cross compiler. Maybe `tarm64osx` as the target machine is not recorded correctly in "~/sandbox/racket-ios/racket/etc/config.rktd", in which case using `--compile-machine tarm64osx` may work around the problem. Or maybe Racket just doesn't complain about a missing cross compiler like it should.

Reply:
> My vague impression ws that W&X is allowed on iOS with the right entitlements and maybe flags like MAP_JIT. Is that just not right?

As far as I understand, WX via `MAP_JIT` is only allowed on iOS using a blessed entitlement that only Apple has access to and which they themselves use in Safari. You can probably add it to a project and get a debug build to work, but that build would be rejected by the App Store review process. It used to be possible to [map the same page twice with different virtual addresses and permissions](https://saagarjha.com/blog/2020/02/23/jailed-just-in-time-compilation-on-ios/), but that was patched in iOS 14. Recently, iOS 14.2 introduced a new capability called "Extended Virtual Addressing" that supposedly allows JIT compilation, but it's undocumented and it appears to require the `get-task-allow` entitlement to work, which is not allowed on the App Store. 

It's unfortunate, but it seems like swapping page protection flags is the only viable option at the moment. 

> More importantly, it can't work right with threads, because one thread might be trying to run some (immobile) code while another thread needs to write on the same pages.

It's definitely not ideal, but I think applications could load all code before launching places (no `dynamic-require`, no `eval`) and avoid futures and still get some of the benefits of threading. Do I have that wrong? Are there more language features that would be limited by this approach?

Thank you for the pointers re. cross-compilation! I'll dig into that tomorrow.

Reply:
With places, I worry that there are cases where an `eval` might show up or code might be lazily unpacked (due to a large module whose outer layers are interpreted), so being sure that everything works is be very difficult. But since allocation will be place-local even in those cases, probably things will work *almost* always.

Reply:
Re. cross-compilation, what do you think of the change in https://github.com/racket/racket/pull/3607/commits/75124e3ed1b55d2bf7e6406e7fb69ef96c4b437f ? Is it too much of a hack? The best fix here would probably be to port the fasl reader in "fasl.c" to Scheme so that it can be patched for cross-compilation like the writer is, but that seems like a significant amount of work. Another approach might be to re-structure `compiler/embed` so that it always uses sources to lay things out, but that won't work if sources aren't available (which might be an OK compromise).

If the changes to `compiler/embed` seem OK, I'll add a change to update the "Inside Racket CS" docs with a section on how to cross compile module images.

Re. threading, I've added a doc note describing the limitations. Alternatively, we could disable places for iOS cross-builds (though my preference is to leave them on despite the limitations).

Reply:
While there is already a second reader for Chez Scheme fasl files that is written in Scheme (it's in "strip.ss", also used for "vfasl.ss"), I don't think it's just a matter of reading the code. Some compile-time code must be run to extract the paths — that is, `eval` is used here. Unless I misunderstand, that means there's just no way to use compiled code for a different platform. I haven't understood every detail of the change for embedding, but it seems like the right idea.

Leaving places enabled seems ok. Unfortunately, it's not necessarily just about places, because some relatively low-level libraries may create threads for other reasons. The recent readline change is an example, although that example is not so relevant to iOS.

I wonder whether the chance for W/X flipping to go bad with threads could be reduced further by flipping only for segments in new enough generations. In a GC, there's an upper bound on the segments that might be modified. In fasl reading and other code-allocation cases, only generation-0 segments can be modified (except vfasl loading, which might be in direct-to-static mode). Or maybe it's so much simpler to work at the chunk level that it's not worth trying to constrain to certain segments.

Reply:
> Some compile-time code must be run to extract the paths — that is, eval is used here. 

Ah! That makes sense and I even went over the code that does this while making the change but did not make this connection. The approach seems more reasonable now. I'll adjust the comments to reflect this.

The change looks much bigger than it is because I had to move some functions around for re-use. The relevant bits are mainly just the ones starting with [this comment](https://github.com/racket/racket/commit/75124e3ed1b55d2bf7e6406e7fb69ef96c4b437f#diff-a40a04285d3ca3db269357d8c60fdbb2464dc302831b3d9079f540a56fc79340R747). Aside from replacing the host code with the target code, the module for that code must also be "prepared" (re-named, submodules potentially removed) in the same ways that the host code was.

> I wonder whether the chance for W/X flipping to go bad with threads could be reduced further by flipping only for segments in new enough generations. In a GC, there's an upper bound on the segments that might be modified. In fasl reading and other code-allocation cases, only generation-0 segments can be modified (except vfasl loading, which might be in direct-to-static mode). Or maybe it's so much simpler to work at the chunk level that it's not worth trying to constrain to certain segments.

I'll take a look at this and report back.

Reply:
What do you think of https://github.com/racket/racket/pull/3607/commits/d84c2eabd68d578626e42ab41c57b48d714ddb00 ? The code could be improved (in particular, I'd want to get rid of `is_unused_seg` or at least add a special-case when a chunk has no used segments), but I wanted to first see if this is along the lines of what you were thinking.

Reply:
I think the changes are on the right track for this idea. And it looks possible to further confine toggling: in cases where toggling is confined to generation 0, that could mean generation 0 for segments that are owned by the current thread (as reflected by the `creator_gc` field in `seginfo`), because it's always enabling writes to just-allocated memory.

I'm still not sure this is worthwhile, since it seems so difficult to make it something you can really reason about and rely on. But maybe it works in practice, and maybe it's worthwhile to keep shrinking the problem surface like this.

Minor: I'd get rid of the `mingen` argument, since it's always 0.

Reply:
Idea for an even further refinement: Starting code write for generation 0 really means the current code allocation segments for the current thread (starts as `base_loc[0][space_code]`) plus any newly allocated segments (via `S_find_more_gc_room`). Ending code write for 0 needs to turn on execution for the current allocation segments plus other newly allocated segments. The `sweep_next` chain in `seginfo`, which currently serves a role only for segments allocated during a collection, could be used to track the latter.

Reply:
I was able to limit flipping to segments belonging to the current thread for generation 0, but had trouble getting the second suggestion to work. I'll give it another go this weekend since I'll have more time then to dig into what exactly is going wrong. 

As a quick stress-test for the two approaches, I set up a project that runs `eval` on a loop inside a place and lets the user `eval` arbitrary code on the main thread. I haven't been able to get it to crash once with the changes to confine toggling, whereas I can easily get it to crash with the coarse-grained approach.

Reply:
A thought on constraining generation-0 writes to active allocation segments: The functions `$set-code-word!`, etc., and `$make-relocation-table!` will be called many times for a given code object, so may not be obvious what "current code allocation segment" means for each individual call. A better strategy for those functions may be to enable writing for just the specific segment containing the address that's an argument to those functions.

For fasl reading, meanwhile, the calls to start and end bracket all allocation and corresponding modification of code, so it makes sense to track the current allocation region there. And for GC, the existing strategy should be ok.

Here are the two remaining problems I see:
 * `__collect_safe` callbacks during `__collect_safe` callouts just can't work. Fortunately, `__collect_safe` functionality (`#:blocking? #t` at the Racket level) is not needed often, but it is needed for the GUI library on Mac OS, and I can imagine that it would be useful for similar reasons on iOS.
 * A function that is compiled and then immediately passed to another Scheme thread runs the risk of becoming non-executable when the original allocating thread allocates more. If a program controls threads and sharing enough, it may be able to avoid doing that, at worst with a `(collect-garbage)` just before passing the new code to the other thread.

Reply:
Thank you for the suggestions! I've added segment hinting in https://github.com/racket/racket/pull/3607/commits/bed4988b29d44091a0bdad50c373e84e8da920d1 (with portability fixes in https://github.com/racket/racket/pull/3607/commits/4b385623333decd976037ce26cc6844ee1a69ce9) and the ability to flip just the current allocation region in https://github.com/racket/racket/pull/3607/commits/75468bc10e2ff0a9df076564734ef0e2766579ff . For hinting, I had to use the effective address rather than the code pointers since the offsets could cause the modifications to be on different segments. For the current region changes, I think I could've avoided adding `sweep_bytes` to `seginfo` and walked the segment up to the forward marker to determine its length, but this approach seemed more clean.

> __collect_safe callbacks during __collect_safe callouts just can't work. Fortunately, __collect_safe functionality (#:blocking? #t at the Racket level) is not needed often, but it is needed for the GUI library on Mac OS, and I can imagine that it would be useful for similar reasons on iOS.

Would it make sense for now to make these fail with an explicit error when `WRITE_XOR_EXECUTE_CODE` is defined?

Reply:
I'm not sure how practical it will be to make those fail. Using `__collect_safe` with a foreign call is not necessarily broken. I guess using `_collect_safe` with a callback could be disabled; although that's not actually a problem when called though a callback without `__collect_safe`, putting `_collect_safe` on a callback declares an intent. But the `WRITE_XOR_EXECUTE_CODE` mode is not curently reflected to the Scheme layer where `foreign-callback` lives. I' inclined to leave this to documentation.